test/libraries/Sign.sol

```diff
@@ -0,0 +1,122 @@
+        bytes data;
+    }
+
+    struct Eip1271Signature {
+        address verifier;
+        bytes signature;
+    }
+
+    /// @dev Encode and sign the order using the provided signing scheme (EIP-712 or EthSign)
+    function sign(
+        Vm vm,
+        Vm.Wallet memory owner,
+        GPv2Order.Data memory order,
+        GPv2Signing.Scheme scheme,
+        bytes32 domainSeparator
+    ) internal returns (Signature memory signature) {
+        bytes32 hash = order.hash(domainSeparator);
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+        if (scheme == GPv2Signing.Scheme.Eip712) {
+            (v, r, s) = vm.sign(owner, hash);
+        } else if (scheme == GPv2Signing.Scheme.EthSign) {
+            (v, r, s) = vm.sign(owner, toEthSignedMessageHash(hash));
+        } else {
+            revert(
+                "Cannot create a signature for the specified signature scheme, only ECDSA-based schemes are supported"
+            );
+        }
+
+        signature.data = abi.encodePacked(r, s, v);
+        signature.scheme = scheme;
+    }
+
+    /// @dev Encode the data used to verify a pre-signed signature
+    function preSign(address owner) internal pure returns (Signature memory) {
+        return Signature(GPv2Signing.Scheme.PreSign, abi.encodePacked(owner));
+    }
+
+    /// @dev Decode the data used to verify a pre-signed signature
+    function toPreSignSignature(Signature memory encodedSignature) internal pure returns (PreSignSignature) {
+        if (encodedSignature.scheme != GPv2Signing.Scheme.PreSign) {
+            revert("Cannot create a signature for the specified signature scheme, only PreSign is supported");
+        }
+
+        address owner;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            owner := shr(96, mload(add(encodedSignature, 0x20)))
+        }
+        return PreSignSignature.wrap(owner);
+    }
+
+    /// @dev Encodes the necessary data required to verify an EIP-1271 signature
+    function sign(EIP1271Verifier verifier, bytes memory signature) internal pure returns (Signature memory) {
+        return Signature(GPv2Signing.Scheme.Eip1271, abi.encodePacked(verifier, signature));
+    }
+
+    /// @dev Decodes the data used to verify an EIP-1271 signature
+    function toEip1271Signature(Signature memory encodedSignature) internal pure returns (Eip1271Signature memory) {
+        if (encodedSignature.scheme != GPv2Signing.Scheme.Eip1271) {
+            revert("Cannot create a signature for the specified signature scheme, only EIP-1271 is supported");
+        }
+
+        address verifier;
+        uint256 length = encodedSignature.data.length - 20;
+        bytes memory signature = encodedSignature.data.slice(20, length);
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly ("memory-safe") {
+            let ptr := add(encodedSignature, 0x40)
+            verifier := shr(96, mload(add(ptr, 0x20)))
+        }
```

Comment:
One more assembly line can be shaved off. :smile:
With the current code, adding a field to the struct `Signature` would also break the assembly code.

```suggestion
        bytes memory signatureData = encodedSignature.data;
        bytes memory signature = signatureData.slice(20, length);

        // solhint-disable-next-line no-inline-assembly
        assembly ("memory-safe") {
            verifier := shr(96, mload(add(signatureData, 0x20)))
        }
```